File: src/main/java/io/github/suitetecsa/sdk/nauta/utils/ExceptionFactoryImpl.java
package io.github.suitetecsa.sdk.nauta.utils;

import io.github.suitetecsa.sdk.nauta.exception.NautaException;

import java.lang.reflect.Constructor;
import java.lang.reflect.InvocationTargetException;

public class ExceptionFactoryImpl<T extends Throwable> implements ExceptionFactory<T> {
    private final Class<T> exceptionClass;

    public ExceptionFactoryImpl(Class<T> exceptionClass) {
        this.exceptionClass = exceptionClass;
    }


    /**
     * Crea una instancia de excepción utilizando la clase de excepción y el mensaje dado.
     *
     * @param message El mensaje de la excepción.
     * @return Una instancia de excepción creada.
     */
    @Override
    public T createException(String message) {
        Constructor<T> constructor;
        try {
            constructor = exceptionClass.getDeclaredConstructor(String.class);
            return constructor.newInstance(message);
        } catch (InstantiationException | IllegalAccessException | InvocationTargetException | NoSuchMethodException e) {
            throw new RuntimeException(e);
        }
    }

    public static class Builder<T extends Exception> {
        private Class<T> exceptionClass;

        public Builder<T> withExceptionClass(Class<T> exceptionClass) {
            this.exceptionClass = exceptionClass;
            return this;
        }

        @SuppressWarnings("unchecked")
        public ExceptionFactoryImpl<T> build() {
            if (exceptionClass == null) exceptionClass = (Class<T>) NautaException.class;
            return new ExceptionFactoryImpl<>(exceptionClass);
        }
    }
}

File: src/main/kotlin/io/github/suitetecsa/sdk/nauta/utils/ExceptionFactoryImpl.kt
package io.github.suitetecsa.sdk.nauta.utils

import io.github.suitetecsa.sdk.nauta.exception.NautaException
import java.lang.reflect.Constructor
import java.lang.reflect.InvocationTargetException

class ExceptionFactoryImpl<T : Throwable>(private val exceptionClass: Class<T>) : ExceptionFactory<T> {
    /**
     * Crea una instancia de excepción utilizando la clase de excepción y el mensaje dado.
     *
     * @param message El mensaje de la excepción.
     * @return Una instancia de excepción creada.
     */
    override fun createException(message: String): T {
        val constructor: Constructor<T>
        try {
            constructor = exceptionClass.getDeclaredConstructor(String::class.java)
            return constructor.newInstance(message)
        } catch (e: InstantiationException) {
            throw RuntimeException(e)
        } catch (e: IllegalAccessException) {
            throw RuntimeException(e)
        } catch (e: InvocationTargetException) {
            throw RuntimeException(e)
        } catch (e: NoSuchMethodException) {
            throw RuntimeException(e)
        }
    }

    class Builder<T : Exception> {
        private var exceptionClass: Class<T>? = null

        fun withExceptionClass(exceptionClass: Class<T>) = apply { this.exceptionClass = exceptionClass }

        fun build(): ExceptionFactoryImpl<T> {
            return ExceptionFactoryImpl(exceptionClass ?: NautaException::class.java as Class<T>)
        }
    }
}
